File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode current = head;
        int length = 0;
        while(current != null) {
            length++;
            current = current.next;
        }
        current = head;
        ListNode newHead = null;
        ListNode tail = null;
        while(length >= k) {
            ListNode first = current;
            ListNode prev = null;
            for (int i = 0; i < k; i++) {
            ListNode nextNode = current.next;
            current.next = prev;
            prev = current;
            current = nextNode;
            }
            if (newHead == null) {
                newHead = prev;
            } else {
                tail.next = prev;
            }
            length -= k;
            tail = first;
        }
        tail.next = current;
        return newHead;
    }
}

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def get_length(node):
            length = 0
            while node:
                length += 1
                node = node.next
            return length

        length = get_length(head)
        new_head, tail = None, None
        current = head

        while length >= k:
            first = current
            prev = None

            for _ in range(k):
                next_node = current.next
                current.next = prev
                prev = current
                current = next_node

            if not new_head:
                new_head = prev
            else:
                tail.next = prev

            tail = first
            length -= k

        if tail:
            tail.next = current

        return new_head or head
